mantidproject/mantid
# Extract computations from ReflectometryMomentumTransfer to a separate algorithm

**Description of work.**

Extracted the calculation of certain statistical quantities related to the beam divergence etc. from  `ReflectometryMomentumTransfer` to a new algorithm, `ReflectometryBeamStatistics` simplifying the ILL's reflectometry reduction workflow. The said quantities are also needed in `ReflectometrySumInQ` so we avoid calculating things twice.

**Report to:** [nobody].

**To test:**

Test data is available in the unit test data directory. Make sure you have the data search directories set up correctly.

Run the test script:
```python
ReflectometryILLPreprocess(
    'ILL/D17/317369',
    OutputWorkspace='direct',
    ForegroundHalfWidth=[2, 3])

ReflectometryILLPreprocess(
    'ILL/D17/317370',
    OutputWorkspace='reflected',
    ForegroundHalfWidth=[3])

def foreground(ws):
    fgd = list()
    run = mtd[ws].run()
    fgd.append(run.getProperty('foreground.first_workspace_index').value)
    fgd.append(run.getProperty('foreground.centre_workspace_index').value)
    fgd.append(run.getProperty('foreground.last_workspace_index').value)
    return fgd

ReflectometryBeamStatistics(
    ReflectedBeamWorkspace='reflected',
    ReflectedForeground=foreground('reflected'),
    DirectBeamWorkspace='direct',
    DirectForeground=foreground('direct'),
    PixelSize=0.001195,
    DetectorResolution=0.0022,
    FirstSlitName='slit2',
    FirstSlitSizeSampleLog='VirtualSlitAxis.s2w_actual_width',
    SecondSlitName='slit3',
    SecondSlitSizeSampleLog='VirtualSlitAxis.s3w_actual_width')
```

Check that the `reflected` workspace now has the following sample log entries: `beam_stats.beam_rms_variation`, `beam_stats.bent_sample`, `beam_stats.first_slit_angular_spread`, `beam_stats.incident_angular_spread`, `beam_stats.sample_waviness`, `beam_stats.second_slit_angular_spread`.

Check that the `direct` workspace now has the `beam_stats.beam_rms_variation` sample log entry.

To test that `ReflectometryMomentumTransfer` still works fine and utilizes the log entries added by `ReflectometryBeamStatistics`, run the following script (uses the workspaces from the script above):
```python
run = mtd['reflected'].run()
first = run.getProperty('foreground.first_workspace_index').value
last = run.getProperty('foreground.last_workspace_index').value
SumSpectra(
    InputWorkspace='reflected',
    OutputWorkspace='reflected-foreground',
    StartWorkspaceIndex=first,
    EndWorkspaceIndex=last)

ReflectometryMomentumTransfer(
    'reflected-foreground',
    OutputWorkspace='R_Q',
    ReflectedForeground=[first, last],
    PixelSize=0.001195,
    DetectorResolution=0.0022,
    ChopperSpeed=1000.,
    ChopperOpening=3.,
    ChopperRadius=0.3,
    ChopperPairDistance=0.4,
    FirstSlitName='slit2',
    FirstSlitSizeSampleLog='VirtualSlitAxis.s2w_actual_width',
    SecondSlitName='slit3',
    SecondSlitSizeSampleLog='VirtualSlitAxis.s3w_actual_width',
    TOFChannelWidth=57.)
```

The script should execute without issues.

Fixes #23622.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
